ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java

```diff
@@ -118,57 +119,49 @@ private void fetchDataFromReadSet() throws MessageFormatException {
+            blobAll.writerIndex((int) (headerFormat.getBlobRecordRelativeOffset() + headerFormat.getBlobRecordSize()));
+          }
         } else {
-          long startTime = SystemTime.getInstance().milliseconds();
           BufferedInputStream bufferedInputStream =
               new BufferedInputStream(new MessageReadSetIndexInputStream(readSet, i, 0),
                   BUFFERED_INPUT_STREAM_BUFFER_SIZE);
-          // read and verify header version
-          byte[] headerVersionBytes = new byte[Version_Field_Size_In_Bytes];
-          bufferedInputStream.read(headerVersionBytes, 0, Version_Field_Size_In_Bytes);
-          short version = ByteBuffer.wrap(headerVersionBytes).getShort();
-          if (!isValidHeaderVersion(version)) {
-            throw new MessageFormatException(
-                "Version not known while reading message - version " + version + ", StoreKey " + readSet.getKeyAt(i),
-                MessageFormatErrorCodes.Unknown_Format_Version);
-          }
-          logger.trace("Calculate offsets, read and verify header version time: {}",
-              SystemTime.getInstance().milliseconds() - startTime);
-
-          // read and verify header
-          startTime = SystemTime.getInstance().milliseconds();
-          byte[] headerBytes = new byte[getHeaderSizeForVersion(version)];
-          bufferedInputStream.read(headerBytes, Version_Field_Size_In_Bytes,
-              headerBytes.length - Version_Field_Size_In_Bytes);
-
-          ByteBuffer header = ByteBuffer.wrap(headerBytes);
-          header.putShort(version);
-          header.rewind();
-          MessageHeader_Format headerFormat = getMessageHeader(version, header);
-          headerFormat.verifyHeader();
-          logger.trace("Calculate offsets, read and verify header time: {}",
-              SystemTime.getInstance().milliseconds() - startTime);
-
-          // read and verify storeKey
-          startTime = SystemTime.getInstance().milliseconds();
-          StoreKey storeKey = storeKeyFactory.getStoreKey(new DataInputStream(bufferedInputStream));
-          if (storeKey.compareTo(readSet.getKeyAt(i)) != 0) {
-            throw new MessageFormatException(
-                "Id mismatch between metadata and store - metadataId " + readSet.getKeyAt(i) + " storeId " + storeKey,
-                MessageFormatErrorCodes.Store_Key_Id_MisMatch);
-          }
-          logger.trace("Calculate offsets, read and verify storeKey time: {}",
-              SystemTime.getInstance().milliseconds() - startTime);
 
-          startTime = SystemTime.getInstance().milliseconds();
+          MessageHeader_Format headerFormat = parseHeaderAndVerifyStoreKey(bufferedInputStream, i);
+
+          long startTime = SystemTime.getInstance().milliseconds();
           if (flag == MessageFormatFlags.BlobProperties) {
             sendInfoList.add(i, new SendInfo(headerFormat.getBlobPropertiesRecordRelativeOffset(),
                 headerFormat.getBlobPropertiesRecordSize()));
```

Comment:
Can the `flag == Blob` case on line 203 be removed now?

Reply:
removed.